File: jar/src/main/java/jline/opt/objectives/SystemResponseTimeConstraint.java
package jline.opt.objectives;

import jline.lang.JobClass;
import jline.opt.results.EvaluationResult;

import java.util.Map;

/** End-to-end (chain-level) response time constraint: SysRespT &lt;= maxValue. */
public class SystemResponseTimeConstraint extends Constraint {

    private final String jobclass;   // null for throughput-weighted aggregate
    private final double maxValue;

    public SystemResponseTimeConstraint(String jobclass, double maxValue, String name) {
        super(name);
        this.jobclass = jobclass;
        this.maxValue = maxValue;
    }

    public SystemResponseTimeConstraint(JobClass jobclass, double maxValue) {
        this(jobclass == null ? null : jobclass.getName(), maxValue, null);
    }

    public SystemResponseTimeConstraint(double maxValue) {
        this((String) null, maxValue, null);
    }

    protected String generateName() {
        if (jobclass != null) {
            return "SysRT_" + jobclass + "_le_" + maxValue;
        }
        return "SysRT_le_" + maxValue;
    }

    public String getJobClass() {
        return jobclass;
    }

    public double getMaxValue() {
        return maxValue;
    }

    public double evaluate(EvaluationResult result, Map<String, Object> variableValues) {
        double actual = (jobclass != null)
                ? result.getSystemResponseTime(jobclass)
                : result.getSystemResponseTime();
        return upperBoundViolation(actual, maxValue);
    }
}

File: matlab/src/opt/+opt/SystemResponseTimeConstraint.m
classdef SystemResponseTimeConstraint < opt.Constraint
    % SystemResponseTimeConstraint  End-to-end response time: SysRespT <= maxValue.
    properties
        jobclass = '';
        maxValue
    end
    methods
        function obj = SystemResponseTimeConstraint(jobclass, maxValue, name)
            if nargin < 3, name = ''; end
            obj@opt.Constraint(name);
            if nargin >= 1 && ~isempty(jobclass)
                obj.jobclass = opt.Constraint.nameOf(jobclass);
            end
            obj.maxValue = maxValue;
        end
        function n = generateName(obj)
            if ~isempty(obj.jobclass)
                n = sprintf('SysRT_%s_le_%g', obj.jobclass, obj.maxValue);
            else
                n = sprintf('SysRT_le_%g', obj.maxValue);
            end
        end
        function v = evaluate(obj, result, ~)
            if ~isempty(obj.jobclass)
                actual = result.getSystemResponseTime(obj.jobclass);
            else
                actual = result.getSystemResponseTime();
            end
            v = opt.Constraint.upperBoundViolation(actual, obj.maxValue);
        end
    end
end
